pkg/agent/proxy/proxier.go

```diff
@@ -303,6 +310,120 @@ func (p *proxier) removeStaleEndpoints(svcPortName k8sproxy.ServicePortName, pro
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *proxier) removeStaleConntrackEntries(svcPortName k8sproxy.ServicePortName, pSvcInfo, svcInfo *types.ServiceInfo, staleEndpoints map[string]k8sproxy.Endpoint) bool {
+	pSvcPort := uint16(pSvcInfo.Port())
+	svcPort := uint16(svcInfo.Port())
+	pNodePort := uint16(pSvcInfo.NodePort())
+	nodePort := uint16(svcInfo.NodePort())
+	pExternalIPStrings := pSvcInfo.ExternalIPStrings()
+	externalIPStrings := svcInfo.ExternalIPStrings()
+	pLoadBalancerIPStrings := pSvcInfo.LoadBalancerIPStrings()
+	loadBalancerIPStrings := svcInfo.LoadBalancerIPStrings()
+
+	staleSvcIPToPort := make(map[string]uint16)
+	// If the port of the Service is changed, delete all conntrack entries related to the previous Service IPs and the
+	// previous Service port. These previous Service IPs includes external IPs, loadBalancer IPs and the ClusterIP.
+	if pSvcPort != svcPort {
+		staleSvcIPToPort[pSvcInfo.ClusterIP().String()] = pSvcPort
+		for _, ip := range pExternalIPStrings {
+			staleSvcIPToPort[ip] = pSvcPort
+		}
+		for _, ip := range pLoadBalancerIPStrings {
+			if ip != "" {
+				staleSvcIPToPort[ip] = pSvcPort
+			}
+		}
+	} else {
+		// If the port of the Service is not changed, delete the conntrack entries related to the stale Service IPs and
+		// the Service port. These stale Service IPs could be from external IPs and loadBalancer IPs.
+		deletedExternalIPs := smallSliceDifference(pExternalIPStrings, externalIPStrings)
+		deletedLoadBalancerIPs := smallSliceDifference(pLoadBalancerIPStrings, loadBalancerIPStrings)
+		for _, ip := range deletedExternalIPs {
+			staleSvcIPToPort[ip] = pSvcPort
+		}
+		for _, ip := range deletedLoadBalancerIPs {
+			staleSvcIPToPort[ip] = pSvcPort
+		}
+	}
+	// If the NodePort of the Service is changed, delete the contrack entries related to the Node IPs and the Service nodePort.
+	if pNodePort != nodePort {
+		for _, nodeIP := range p.nodePortAddresses {
+			staleSvcIPToPort[nodeIP.String()] = pNodePort
+		}
+	}
+	// Delete the conntrack entries due to the change of the Service.
+	for svcIPStr, port := range staleSvcIPToPort {
+		svcIP := net.ParseIP(svcIPStr)
+		if err := p.routeClient.ClearConntrackEntryForService(svcIP, port, nil, pSvcInfo.OFProtocol); err != nil {
+			klog.ErrorS(err, "Error when removing conntrack for Service", "ServicePortName", svcPortName, "ServiceIP", svcIP, "ServicePort", port)
+			return false
+		}
+	}
+
+	existingSvcIPToPort := make(map[string]uint16)
+	existingSvcIPToPort[svcInfo.ClusterIP().String()] = svcPort
+	// Get all existing Service IPs.
+	existingExternalIPs := smallSliceSame(pExternalIPStrings, externalIPStrings)
+	existingLoadBalancerIPs := smallSliceSame(pLoadBalancerIPStrings, loadBalancerIPStrings)
+	for _, ip := range existingExternalIPs {
+		existingSvcIPToPort[ip] = svcPort
+	}
+	for _, ip := range existingLoadBalancerIPs {
+		existingSvcIPToPort[ip] = svcPort
+	}
```

Comment:
Are't them already cleaned up if port changes?

Reply:
You are right, I will update later.